### PR TITLE
Plugin to use elasticsearch as a batch sink

### PR DESCRIPTION
**Pull request by ataylor4**
This feature will allow users to sink data to elasticsearch, provided an elasticsearch server is running.

http://builds.cask.co/browse/CDAP-DUT2220-2

### CODE REVIEW

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/sink/ElasticsearchSink.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.io.Writable;
+import org.apache.hadoop.mapreduce.Job;
+import org.elasticsearch.hadoop.mr.EsOutputFormat;
+
+import javax.annotation.Nullable;
+
+/**
+ * An ElasticsearchSink
```
Comment (gokulavasan) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/sink/ElasticsearchSink.java]: Some more doc?

File: cdap-app-fabric/pom.xml [hunk 2]
```diff
@@ -122,6 +122,11 @@
       <groupId>org.apache.spark</groupId>
       <artifactId>spark-core_2.10</artifactId>
     </dependency>
+    <dependency>
+      <groupId>org.elasticsearch</groupId>
+      <artifactId>elasticsearch-hadoop-mr</artifactId>
+      <version>2.1.0</version>
```
Comment (gokulavasan) [inline cdap-app-fabric/pom.xml]: do we need this version here? I think the parent pom has the version in the dependency management?

Comment (gokulavasan) [inline cdap-app-fabric/pom.xml]: Hmm, why do we need this dependency here?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/pom.xml [hunk 3]
```diff
@@ -215,6 +215,27 @@
...
+    </dependency>
+    <dependency>
+      <groupId>it.unimi.dsi</groupId>
+      <artifactId>fastutil</artifactId>
+      <version>${fastutil.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>commons-httpclient</groupId>
+      <artifactId>commons-httpclient</artifactId>
+      <version>3.1</version>
```
Comment (gokulavasan) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/pom.xml]: make this a variable for reuse?

Comment (gokulavasan) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/pom.xml]: also add a comment may be to say why we added this dependency?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/RecordUtils.java [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+import java.util.IdentityHashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Utility class for converting {@link StructuredRecord} to and from json.
+ */
+public final class RecordUtils {
```
Comment (gokulavasan) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/RecordUtils.java]: We should probably add some unit tests for this class?

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLESSinkTest.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+  @Before
+  public void beforeTest() throws Exception {
+    ImmutableSettings.Builder elasticsearchSettings = ImmutableSettings.settingsBuilder()
+      .put("path.data", tmpFolder.newFolder("data"));
+    node = nodeBuilder().settings(elasticsearchSettings.build()).client(false).node();
+    client = node.client();
+  }
+
+  @After
+  public void afterTest() { node.close();
```
Comment (jayfjin) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLESSinkTest.java]: move to next line

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/RecordUtils.java [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+    }
+
+    if (Map.class.isAssignableFrom(cls)) {
+      return Schema.Type.MAP;
+    }
+
+    throw new IOException("Unsupported type found in StructuredRecord: " + cls);
+  }
+
+  private RecordUtils() {
```
Comment (jayfjin) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/RecordUtils.java]: //no-op

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLESSinkTest.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+import static org.elasticsearch.index.query.QueryBuilders.matchQuery;
+import static org.elasticsearch.node.NodeBuilder.nodeBuilder;
+
+/**
+ * <p>
+ *  Unit test for {@link ElasticsearchSink} ETL realtime source class.
```
Comment (rohitsinha54) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLESSinkTest.java]: realtime ?

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLESSinkTest.java [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+      List<ETLStage> transforms = Lists.newArrayList();
+      ETLBatchConfig etlConfig = new ETLBatchConfig("* * * * *", source, sink, transforms);
+      Id.Adapter adapterId = Id.Adapter.from(NAMESPACE, "esSinkTest");
+      AdapterConfig adapterConfig = new AdapterConfig("", TEMPLATE_ID.getId(), GSON.toJsonTree(etlConfig));
+      AdapterManager manager = createAdapter(adapterId, adapterConfig);
+
+      manager.start();
+      manager.waitForOneRunToFinish(5, TimeUnit.MINUTES);
+      manager.stop();
+      SearchResponse searchResponse = client.prepareSearch("test").execute().actionGet();
```
Comment (rohitsinha54) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLESSinkTest.java]: Why not use assertEquals ?

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLESSinkTest.java [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+import co.cask.cdap.template.etl.common.ETLStage;
+import co.cask.cdap.template.etl.common.Properties;
+import co.cask.cdap.test.AdapterManager;
+import co.cask.cdap.test.SlowTests;
+import co.cask.cdap.test.StreamManager;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Lists;
+import com.google.gson.Gson;
+import org.elasticsearch.action.admin.indices.create.CreateIndexRequest;
+import org.elasticsearch.action.admin.indices.create.CreateIndexResponse;
```
Comment (rohitsinha54) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLESSinkTest.java]: Request and Response are unused imports here.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/sink/ElasticsearchSink.java [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+import co.cask.cdap.template.etl.api.batch.BatchSinkContext;
+import co.cask.cdap.template.etl.common.Properties;
+import co.cask.cdap.template.etl.common.RecordUtils;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.Text;
+import org.apache.hadoop.io.Writable;
+import org.apache.hadoop.mapreduce.Job;
+import org.elasticsearch.hadoop.mr.EsOutputFormat;
+
+import javax.annotation.Nullable;
```
Comment (rohitsinha54) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/sink/ElasticsearchSink.java]: Unused

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/RecordUtils.java (hunk 4 above)
Comment (rohitsinha54) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/RecordUtils.java]: The classname is sounds generic to be Utility class for Records and the class documentation is specific. If you plan to add support for more conversion then change the doc or else the classname.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/RecordUtils.java [hunk 11]
```diff
@@ -0,0 +1,438 @@
...
+    reader.beginArray();
+    while (reader.peek() != JsonToken.END_ARRAY) {
+      bytes.add((byte) reader.nextInt());
+    }
+    reader.endArray();
+    return bytes.toByteArray();
+  }
+
+  private static List<Object> readArray(JsonReader reader, Schema elementSchema) throws IOException {
+    List<Object> result = Lists.newArrayList();
```
Comment (jayfjin) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/RecordUtils.java]: new ArrayList();

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/RecordUtils.java [hunk 12]
```diff
@@ -0,0 +1,438 @@
...
+
+  private static Map<Object, Object> readMap(JsonReader reader,
+                                             Map.Entry<Schema, Schema> mapSchema) throws IOException {
+    Schema keySchema = mapSchema.getKey();
+    if (!keySchema.isCompatible(Schema.of(Schema.Type.STRING))) {
+      throw new IOException("Complex key type not supported: " + keySchema);
+    }
+
+    Schema valueSchema = mapSchema.getValue();
+    Map<Object, Object> result = Maps.newHashMap();
```
Comment (jayfjin) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/RecordUtils.java]: new HashMap();

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLESSinkTest.java [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+      .build());
+
+    try {
+      ETLStage sink = new ETLStage("Elasticsearch",
+                                   ImmutableMap.of(Properties.Elasticsearch.HOST, "localhost:9200",
+                                                   Properties.Elasticsearch.INDEX_NAME, "test",
+                                                   Properties.Elasticsearch.TYPE_NAME, "testing",
+                                                   Properties.Elasticsearch.ID_FIELD, "ticker"
+                                   ));
+      List<ETLStage> transforms = Lists.newArrayList();
```
Comment (jayfjin) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLESSinkTest.java]: new ArrayList();

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLESSinkTest.java (hunk 7 above)
Comment (jayfjin) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLESSinkTest.java]: and source?

File: cdap-ui/templates/ETLBatch/Elasticsearch.json [hunk 14]
```diff
@@ -0,0 +1,45 @@
+{
+  "id": "ElasticsearchSink",
+  "groups" : {
+    "position": [ "group1"],
```
Comment (jayfjin) [inline cdap-ui/templates/ETLBatch/Elasticsearch.json]: space after last element, for consistent style

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/sink/ElasticsearchSink.java [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+
+  private static final String TYPE_DESC = "The name of the type where the data will be stored";
+
+  private static final String ID_DESC = "The id field that will determine the id for the document. " +
+    "It should match a fieldname in the structured record of the input";
+
+  private static final String HOST_DESC = "The hostname and port for an elasticsearch node. e.g. localhost:9300";
+
+  /**
+   * Config class for RealtimeTableSink
```
Comment (jayfjin) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/sink/ElasticsearchSink.java]: incorrect javadoc

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/sink/ElasticsearchSink.java [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+   */
+  public static class ESConfig extends PluginConfig {
```
Comment (jayfjin) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/sink/ElasticsearchSink.java]: also can we move this class to the bottom of the class? it's sandwiched awkwardly between final variables

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/sink/ElasticsearchSink.java [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+/**
+ * An ElasticsearchSink
+ */
+@Plugin(type = "sink")
+@Name("Elasticsearch")
+@Description("CDAP Elasticsearch Batch Sink")
+public class ElasticsearchSink extends BatchSink<StructuredRecord, Writable, Writable> {
+  private static final String INDEX_DESC = "The name of the index where the data will be stored. " +
+    "The index should already exist and be configured.";
+
```
Comment (jayfjin) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/sink/ElasticsearchSink.java]: are these newlines necessary?

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/spark/SparkAppUsingFileSet.java [hunk 18]
```diff
@@ -95,7 +95,7 @@ public void configure() {
   public static final class ScalaCharCount extends AbstractSpark {
     @Override
     public void configure() {
-      setMainClass(ScalaFileCountProgram.class);
+      //setMainClass(ScalaFileCountProgram.class);
```
Comment (gokulavasan) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/spark/SparkAppUsingFileSet.java]: Is this an intended change? If yes, what is this for?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/sink/ElasticsearchSink.java [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+import org.apache.hadoop.io.Writable;
+import org.apache.hadoop.mapreduce.Job;
+import org.elasticsearch.hadoop.mr.EsOutputFormat;
+
+/**
+ * A {@link BatchSink} that writes data to a Elasticsearch.
+ * <p/>
+ * This {@link ElasticsearchSink} takes a {@link StructuredRecord} in,
+ * converts it to a json per {@link StructuredRecordStringConversion},
+ * and writes it to the Elasticserach server.
```
Comment (anwar6953) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/sink/ElasticsearchSink.java]: misspelling in ElasticSearch

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/sink/ElasticsearchSink.java [hunk 20]
```diff
@@ -0,0 +1,117 @@
...
+ * <p/>
+ * If the Elasticserach index does not exist, it will be created using the default properties
```
Comment (anwar6953) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/sink/ElasticsearchSink.java]: misspelling in ElasticSearch

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/sink/ElasticsearchSink.java [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+ * specified by Elasticsearch. See more information at
+ * https://www.elastic.co/guide/en/elasticsearch/guide/current/_index_settings.html.
+ * <p/>
+ */
+@Plugin(type = "sink")
+@Name("Elasticsearch")
+@Description("CDAP Elasticsearch Batch Sink")
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/sink/ElasticsearchSink.java]: Should probably add a lot more description. Maybe some stuff from this class' javadoc.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/StructuredRecordStringConversion.java [hunk 22]
```diff
@@ -0,0 +1,439 @@
...
+import java.util.IdentityHashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Utility class for converting {@link StructuredRecord} to and from json.
+ */
+public final class StructuredRecordStringConversion {
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/StructuredRecordStringConversion.java]: s/Conversion/Converter/

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/StructuredRecordStringConversion.java [hunk 23]
```diff
@@ -0,0 +1,439 @@
...
+
+  // Known Java type to schema type mapping
+  // Doesn't have map and array as those need to use instanceof to check
+  private static final Map<Class<?>, Schema.Type> TYPE_TO_SCHEMA = new IdentityHashMap<Class<?>, Schema.Type>(
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/StructuredRecordStringConversion.java]: Can this just be ``new IdentityHashMap<>``

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/StructuredRecordStringConversion.java [hunk 24]
```diff
@@ -0,0 +1,439 @@
...
+      .put(Float.class, Schema.Type.FLOAT)
+      .put(Double.class, Schema.Type.DOUBLE)
+      .put(String.class, Schema.Type.STRING)
+      .put(ByteBuffer.class, Schema.Type.BYTES)
+      .put(byte[].class, Schema.Type.BYTES)
+      .put(StructuredRecord.class, Schema.Type.RECORD)
+      .build()
+  );
+
+  private static final EnumMap<Schema.Type, JsonToken> SCHEMA_TO_JSON_TYPE = new EnumMap<Schema.Type, JsonToken>(
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/StructuredRecordStringConversion.java]: can this just be ``new EnumMap<>``?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/StructuredRecordStringConversion.java [hunk 25]
```diff
@@ -0,0 +1,439 @@
...
+   * Converts a {@link StructuredRecord} to a json string.
+   */
+  public static String toJsonString(StructuredRecord record) {
+    StringWriter strWriter = new StringWriter();
+    JsonWriter writer = new JsonWriter(strWriter);
+    try {
+      writeJson(writer, record.getSchema(), record);
+      return strWriter.toString();
+    } catch (IOException e) {
+      throw Throwables.propagate(e);
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/StructuredRecordStringConversion.java]: should we throw checked exceptions here, and leave it to the caller to handle them?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/StructuredRecordStringConversion.java [hunk 26]
```diff
@@ -0,0 +1,439 @@
...
+
+  /**
+   * Converts a json string to a {@link StructuredRecord} based on the schema.
+   */
+  public static StructuredRecord fromJsonString(String json, Schema schema) {
+    JsonReader reader = new JsonReader(new StringReader(json));
+    try {
+      return (StructuredRecord) readJson(reader, schema);
+    } catch (IOException e) {
+      throw Throwables.propagate(e);
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/StructuredRecordStringConversion.java]: same. checked exceptions?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/StructuredRecordStringConversion.java [hunk 27]
```diff
@@ -0,0 +1,439 @@
...
+  public static String toJsonString(StructuredRecord record) {
+    StringWriter strWriter = new StringWriter();
+    JsonWriter writer = new JsonWriter(strWriter);
+    try {
+      writeJson(writer, record.getSchema(), record);
+      return strWriter.toString();
+    } catch (IOException e) {
+      throw Throwables.propagate(e);
+    } finally {
+      Closeables.closeQuietly(writer);
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/StructuredRecordStringConversion.java]: can we just use ``writer.close``?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/StructuredRecordStringConversion.java [hunk 28]
```diff
@@ -0,0 +1,439 @@
...
+   * Converts a json string to a {@link StructuredRecord} based on the schema.
+   */
+  public static StructuredRecord fromJsonString(String json, Schema schema) {
+    JsonReader reader = new JsonReader(new StringReader(json));
+    try {
+      return (StructuredRecord) readJson(reader, schema);
+    } catch (IOException e) {
+      throw Throwables.propagate(e);
+    } finally {
+      Closeables.closeQuietly(reader);
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/StructuredRecordStringConversion.java]: ``reader.close()``?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/StructuredRecordStringConversion.java [hunk 29]
```diff
@@ -0,0 +1,439 @@
...
+    } else {
+      throw new IOException("Expects either ByteBuffer or byte[]. Got " + value.getClass());
+    }
+  }
+
+  private static void writeBytes(JsonWriter writer, ByteBuffer buffer) throws IOException {
+    if (buffer.hasArray()) {
+      writeBytes(writer, buffer.array(), buffer.arrayOffset() + buffer.position(), buffer.remaining());
+    } else {
+      byte[] buf = new byte[buffer.remaining()];
```
Comment (anwar6953) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/StructuredRecordStringConversion.java]: Should be able to simplify this by just using `Bytes#getBytes(ByteBuffer)`, right?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/test/java/co/cask/cdap/template/etl/common/StructuredRecordStringConversionTest.java [hunk 30]
```diff
@@ -0,0 +1,87 @@
...
+@SuppressWarnings("unchecked")
+public class StructuredRecordStringConversionTest {
+  Schema schema;
+
+  @Test
+  public void checkConversion() {
+    StructuredRecord initial = getStructuredRecord();
+    String jsonOfRecord = StructuredRecordStringConversion.toJsonString(initial);
+    StructuredRecord recordOfJson = StructuredRecordStringConversion.fromJsonString(jsonOfRecord, schema);
+    areRecordsEqual(initial, recordOfJson);
```
Comment (anwar6953) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/test/java/co/cask/cdap/template/etl/common/StructuredRecordStringConversionTest.java]: the 'are' seems like it returns a boolean.
Perhaps rename to `assertRecordsEqual`

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLESSinkTest.java [hunk 31]
```diff
@@ -0,0 +1,139 @@
...
+      Id.Adapter adapterId = Id.Adapter.from(NAMESPACE, "esSinkTest");
+      AdapterConfig adapterConfig = new AdapterConfig("", TEMPLATE_ID.getId(), GSON.toJsonTree(etlConfig));
+      AdapterManager manager = createAdapter(adapterId, adapterConfig);
+
+      manager.start();
+      manager.waitForOneRunToFinish(5, TimeUnit.MINUTES);
+      manager.stop();
+
+      SearchResponse searchResponse = client.prepareSearch("test").execute().actionGet();
+      Assert.assertEquals(searchResponse.getHits().getTotalHits(), 2);
```
Comment (jayfjin) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/template/etl/batch/ETLESSinkTest.java]: expected value should come first

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/StructuredRecordStringConversion.java [hunk 32]
```diff
@@ -0,0 +1,439 @@
...
+      Object key = convertKey(reader.nextName(), keySchema.getType());
+      result.put(key, readJson(reader, valueSchema));
+    }
+    reader.endObject();
+
+    return result;
+  }
+
+  private static Object convertKey(String key, Schema.Type type) throws IOException {
+    if (type == Schema.Type.STRING) {
```
Comment (jayfjin) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/StructuredRecordStringConversion.java]: is switch case better here?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/test/java/co/cask/cdap/template/etl/common/StructuredRecordStringConversionTest.java [hunk 33]
```diff
@@ -0,0 +1,87 @@
...
+      .set("time", System.currentTimeMillis())
+      .set("binary", "Bob".getBytes(Charsets.UTF_8))
+      .set("innerRecord", innerRecordBuilder.build());
+
+    return recordBuilder.build();
+  }
+
+  private void areRecordsEqual(StructuredRecord one, StructuredRecord two) {
+    Assert.assertTrue(one.getSchema().getRecordName().equals(two.getSchema().getRecordName()));
+    Assert.assertTrue(one.getSchema().getFields().size() == two.getSchema().getFields().size());
```
Comment (jayfjin) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/test/java/co/cask/cdap/template/etl/common/StructuredRecordStringConversionTest.java]: assertEquals?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/sink/ElasticsearchSink.java [hunk 34]
```diff
@@ -0,0 +1,119 @@
...
+
+  private String getResource() {
+    return String.format("%s/%s", config.index, config.type);
+  }
+
+  @Override
+  public void prepareRun(BatchSinkContext context) {
+    Job job = context.getHadoopJob();
+    Configuration conf = job.getConfiguration();
+    conf.setBoolean("mapred.map.tasks.speculative.execution", false);
```
Comment (gokulavasan) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/sink/ElasticsearchSink.java]: Should be instead do ``job.setSpeculativeExecution(false)`` ?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/StructuredRecordStringConverter.java [hunk 35]
```diff
@@ -0,0 +1,430 @@
...
+    for (Schema schema : unionSchema.getUnionSchemas()) {
+      if (SCHEMA_TO_JSON_TYPE.get(schema.getType()) == token) {
+        return readJson(reader, schema);
+      }
+    }
+
+    throw new IOException("No matching schema found for union type: " + unionSchema);
+  }
+
+
```
Comment (jayfjin) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/StructuredRecordStringConverter.java]: extra newline

File: cdap-app-templates/cdap-etl/cdap-etl-lib/pom.xml [hunk 36]
```diff
@@ -215,6 +215,28 @@
...
+      <groupId>org.elasticsearch</groupId>
+      <artifactId>elasticsearch-hadoop-mr</artifactId>
+      <version>${es-hadoop.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>it.unimi.dsi</groupId>
+      <artifactId>fastutil</artifactId>
+      <version>${fastutil.version}</version>
+    </dependency>
+    <!-- elasticsearch-hadoop-mr includes a class, StringUtils, which requires commons-httpclient-->
```
Comment (chtyim) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/pom.xml]: StringUtils uses HttpClient? Not blaming you, but you learn how bad design means.

**Comment by chtyim** [comment]
LGTM. can you squashes all commits into one? Having bunch of "WIP" as commit message is not good.

**Comment by ataylor4** [comment]
Combined commits in PR: https://github.com/caskdata/cdap/pull/3282 and merged that PR

**Pull request closed without merging**